File: 2020/rust/src/main.rs
#[macro_use] extern crate lazy_static;

extern crate elapsed;
extern crate chrono;
extern crate itertools;
extern crate regex;

use std::process::exit;
use std::env;
use std::fs;
use std::io::{self, Read};

use chrono::prelude::*;

pub mod reportrepair;
pub mod passwordphilosophy;
pub mod toboggantrajectory;
pub mod passportprocessing;
pub mod binaryboarding;
pub mod customcustoms;
pub mod handyhaversacks;
pub mod handheldhalting;
pub mod encodingerror;
pub mod adapterarray;
pub mod seatingsystem;
pub mod rainrisk;
pub mod shuttlesearch;
pub mod dockingdata;
pub mod rambunctiousrecitation;
pub mod tickettranslation;

fn usage() {
    println!("usage: adventofcode <day> [-|input file]");
}

const YEAR: i32 = 2020;

fn guess_day() -> Result<u8, i32> {
    let now: DateTime<Local> = Local::now();
    if now.year() != YEAR || now.month() != 12 || now.day() > 25 {
        Err(2)
    } else {
        let unlock = Local.ymd(now.year(), now.month(), now.day())
            .and_hms_milli(10, 30, 0, 0);

        if now > unlock {
            Ok(now.day() as u8)
        } else {
            Ok((now.day() - 1) as u8)
        }
    }
}

fn get_data(day: u8) -> io::Result<String> {
    let mut args: Vec<String> = env::args().collect();
    if args.len() > 2 {
        if args[2] == "-" {
            let mut buffer = String::new();
            let stdin = io::stdin();
            let mut handle = stdin.lock();

            handle.read_to_string(&mut buffer)
                .expect("error reading input data");
            Ok(buffer)
        } else if fs::metadata(&args[2]).is_ok() {
            fs::read_to_string(&args[2])
        } else {
            Ok(args.remove(2))
        }
    } else {
        let filename = format!("../inputs/input{:02}.txt", day);
        fs::read_to_string(filename)
    }
}

fn run(day: u8, data: String) {

    match day {
        1  => { reportrepair::solve(data);           }
        2  => { passwordphilosophy::solve(data);     }
        3  => { toboggantrajectory::solve(data);     }
        4  => { passportprocessing::solve(data);     }
        5  => { binaryboarding::solve(data);         }
        6  => { customcustoms::solve(data);          }
        7  => { handyhaversacks::solve(data);        }
        8  => { handheldhalting::solve(data);        }
        9  => { encodingerror::solve(data);          }
        10 => { adapterarray::solve(data);           }
        11 => { seatingsystem::solve(data);          }
        12 => { rainrisk::solve(data);               }
        13 => { shuttlesearch::solve(data);          }
        14 => { dockingdata::solve(data);            }
        15 => { rambunctiousrecitation::solve(data); }
        16 => { tickettranslation::solve(data);      }
        _  => { println!("no solver for day {} yet.", day); }
    }
}

fn main() {
    let args: Vec<String> = env::args().collect();

    let day: u8 = if args.len() > 1 {
        match args[1].trim().parse() {
            Ok(num) => num,
            Err(_) => { usage(); exit(2); }
        }
    } else {
        match guess_day() {
            Ok(day) => day,
            Err(code) => { usage(); exit(code); }
        }
    };

    let data = match get_data(day) {
        Ok(contents) => contents.trim().to_string(),
        Err(_) => {
            println!("input not found for day {}", day);
            usage(); exit(5);
        }
    };

    let (elapsed, _) = elapsed::measure_time(|| {
        run(day, data);
    });
    println!("elapsed = {}", elapsed);
}

File: 2020/haskell/app/Main.hs
module Main where

import System.Directory       (doesFileExist)
import System.Environment     (getArgs)
import System.CPUTime         (getCPUTime)
import Text.Printf            (printf)
import Data.Time.Clock        (getCurrentTime)
import Data.Time.Calendar     (toGregorian)
import Data.Time.LocalTime
import Control.Applicative    (liftA)
import Control.Monad.IO.Class (MonadIO(liftIO))


import AdventOfCode
import ReportRepair       (day01)
import PasswordPhilosophy (day02)
import TobogganTrajectory (day03)
import PassportProcessing (day04)
import BinaryBoarding     (day05)
import CustomCustoms      (day06)
import HandyHaversacks    (day07)
import HandheldHalting    (day08)
import EncodingError      (day09)
import AdapterArray       (day10)

import RainRisk           (day12)
import ShuttleSearch      (day13)

loadFile :: FilePath -> IO String
loadFile filename = do
  fileExists <- doesFileExist filename
  if fileExists
    then readFile filename
    else errorWithoutStackTrace (printf "Input file " ++ filename ++ " not found")


getInput :: Int -> IO String
getInput day = do
  args <- getArgs
  if (length args) > 1 then loadFile (last args)
    else
    loadFile (printf "../inputs/input%02d.txt" day)


checkUnlocked :: IO Bool
checkUnlocked = do
  let now = getCurrentTime >>= return . localTimeOfDay . (utcToLocalTime (TimeZone 330 False "IST"))
  hours <- (liftA todHour) now
  minutes <- (liftA todMin) now
  return ((hours * 60 + minutes) > 630)


guessDay :: IO Int
guessDay = do
  (year, month, day) <- getCurrentTime >>= return . toGregorian . localDay
                        . (utcToLocalTime (TimeZone 330 False "IST"))
  if year /= 2020 || month /= 12 || day > 25
    then errorWithoutStackTrace "AoC 2020 not currently running, day must be provided."
    else return day


getDay :: IO Int
getDay = do
  args <- getArgs
  case args of
    (day:_) -> return (read day)
    [] -> do
      day <- guessDay
      unlocked <- checkUnlocked
      let currentDay = if unlocked then day else (pred day)
      return currentDay


formatElapsedTime :: Integral t => t -> String
formatElapsedTime elapsed = let t :: Double
                                t = fromIntegral elapsed in
                              formatTime t ["ps", "ns", "us", "ms", "s"]
  where formatTime t units
          | t >= 1000.0 = formatTime (t / 1000.0) (tail units)
          | otherwise   = printf "Elapsed: %4.2f %s" t (head units)

-- solve calls the correct solver after parsing the input into the
-- expected format
solve :: Int -> String -> IO ()
solve  1 = day01 . parseInputNumbers
solve  2 = day02 . lines
solve  3 = day03 . lines
solve  4 = day04
solve  5 = day05 . lines
solve  6 = day06
solve  7 = day07 . lines
solve  8 = day08 . lines
solve  9 = day09 . parseInputNumbers
solve 10 = day10 . parseInputNumbers
solve 12 = day12 . lines
solve 13 = day13 . lines
solve n = error (printf "No solver for day %d yet.\n" n)


main :: IO ()
main = do
  day <- getDay
  input <- getInput day
  start <- liftIO getCPUTime
  res <- solve day (trim input)
  end <- liftIO getCPUTime
  putStrLn $ formatElapsedTime (end - start)
  return res
